### ISSUE TITLE
Memory layout (shared/global/local, AOSOA...) for intermediate wavefunctions in ME calculations

### ISSUE THREAD

**Issue by roiser** [madgraph5/madgraph4gpu#7]
eemumu_AV/master

AV should check for bugs in global memory implementation (seems not possible to exhaust it with eemumu…)

**Comment by valassi** [comment]
This is ongoing. I fixed the global memory issue in the meantime: https://github.com/roiser/madgraph4gpu/commit/649bdc472077db51d81a72a065d0c22b94504e5e

But I still get much worse performance than with local memory or shared memory. It is even worse than before, because now I can use 16384 blocks and the performance penalty of global vs local is much worse at 16384 clocks than it was at 256.

I would say that this remains to be better undesrtoood and profiled

**Comment by valassi** [comment]
I have closed the AOS/SOA issue #16 fo rthe memory layout of momenta (and random numbers). Now the access to momenta in memory seems well coalesced... and this does not even seem to have a big impact.

I have renamed this issue #7 as "Memory layout (shared/global/local, AOSOA...) for intermediate wavefunctions in ME calculations". It may still be that optimising those memory structures like w[5][6] will have an impact.

In other words, the memory of the input data (momenta) to signmakin and of output data (MEs) seems now under control, and maybe the layout is not that relevant. We should rather focus on the memory layout inside the sigmakin calculation, ie nside the ixxx, oxxx and FV functions.

**Comment by valassi** [comment]
I repeat here what I wrote in issue #26 about maxregcount.

I think that I got this whole idea of local/global/shared wrong, because I had not understood two things

- one, that 'registers' are vector registers, and that warps effectively do "simd" n those vector registers (so even the 'local' w[5][6] is already vectorizxed), and there is no need to vectorize them manually as in c++ on a cpu
- two, that what i read about 'local storage is thread-local global storage' is also in part a misconception, because those stack variables like w[5][6] stay on the registers as long as possible, and only spill out to "local" memory (i.e. thread-local global memory) if there is no space left in the registers.

It seems much better therefore to concentrate on the 'local' coding with w[5][6] and drop the shared and global implementations, which are complex and have worse performance. For instance
- In issue #26 I investigated the use of maxregcount. This does improve things, but needs careful fine tuning. Essentially, now with the latest AOSOA structure (issue #16) we are doing minimal usage of the memory bandwidth. Setting a limit on the number of registers forces the spillover from registries to "local" memory, which in any case is not an issue as we have plenty of unused bandwidth. This allows increasing occupancy by reducing the number of registries.
- In issue #23, I wonder if moving some physics constants from constant memory to hardcoded executables would also reduce the number of registries.
- We should investigate individual operations like sqrt (issue #15)
- Clearly using float instead of double would not only allow the use of FP32, but would also reduce the number of registers (issue #5)
- I wonder if using a better complex implementation (issue #6) with home-made wrapping of doubles would also decrease register usage?
- And we should also consider splitting kernels into smaller chunks, for instance using cuda graphs (issue #12)

As far as this work on local vs global vcs storage is concerned, I would say that
- The shared implementation, as it is now for w[5][6], has no future. It has a 35% throughpout penalty and it already uses 15k bytes of shared memory with 32 threads, when there are only 64k (or maybe 48k?). It is already impossible to use 128 threads. With more complex processes like ggttgg. it would simply explode. This is not a useful way for us to use shared memory.
- The global memory implementation has a huge hit, like 70%. As discussed in issue #26, it seems much better to let the system handlk the spill of some memory to "local" (thread-local global) memory, rather than do it ourselves.
- The only reason why we may need to resurrect some of these global memory studies in the future is if we do split the sigmakin into smaller chunks. In that case we need to store intermediate results (like the w[5][6]...) into global memory (NOT shared memory) to move it across kernels. But it would need a much more profound reengineering of the kernels.

In conclusion, I woudl simply remove the present global and shared implementations. THis will clean up tyhe code.

Before doing that, I will still post a few graphs from profiling of shared and global below.

**Comment by valassi** [comment]
First, compare SHARED to "LOCAL".

It uses memory more, but SM usage decreases (unlike the case of maxregcount 128, where using more the memory allows a better usage of SM too). 
![image](https://user-images.githubusercontent.com/3473550/90527252-9eccd700-e171-11ea-9125-b4f28cf50b5a.png)

The roofline is not very interesting. Actually the point moves to the right, not to the left (arithm,etic intensity increases). Actually FP64 decreases, while the LSU load store unit increases a lot
![image](https://user-images.githubusercontent.com/3473550/90527442-d6d41a00-e171-11ea-9a8f-23e34d6b9bdd.png)

The memory plot shows that shared memory is used, which was not the case before. There are also wuit ea few bank conflicts (I am not sure if they are many or if that's an issue)
![image](https://user-images.githubusercontent.com/3473550/90527618-071bb880-e172-11ea-89d0-a1c24b0b1b9d.png)

On the stall plot, it seems that the "useful" stalls like math pipe throttle decrease a lot
![image](https://user-images.githubusercontent.com/3473550/90527749-2f0b1c00-e172-11ea-8840-f6cdb54a3f1f.png)

The mix of instructions changes a lot. But I would say that there a lot of new, not essential operations that are added. So we do more computations maybe, but not in useful things?
![image](https://user-images.githubusercontent.com/3473550/90527936-5f52ba80-e172-11ea-9059-fca71cb3ab4e.png)

Finally, the occupancy plot. This is probably one of the most useful plots here. All this work was to reduce the number of registers, and we do reduce them a bit from 176 to 164. But, because we are using shared memory, this does NOT translate into a higher occupancy, actually the overall occupancy is lower. (I am not sure what is involved, but this is what the bottom plot syas)
![image](https://user-images.githubusercontent.com/3473550/90528232-b062ae80-e172-11ea-927a-fd6a564a58bb.png)

So, in conclusion this usage of shared memory done this way is useless.

**Comment by valassi** [comment]
About global memory. This looks a bit like the plot for maxreg=64. We have increased memory usage so much, that SM decreases and we have become memory bound
![image](https://user-images.githubusercontent.com/3473550/90528471-f15ac300-e172-11ea-9d33-7df49c28af66.png)

And again the roofline says it all, we are now on the diagonal
![image](https://user-images.githubusercontent.com/3473550/90528567-10595500-e173-11ea-9900-c0b5ab781c9f.png)

Not surprisingly, memory usage has increased enormously
![image](https://user-images.githubusercontent.com/3473550/90528660-3121aa80-e173-11ea-9735-8d5dbfa47e6a.png)

And the warp plots are interesting, though I do not understand all of it. We actually have slightly higher active warps per scheduler, but much fewer issued warps per scheduler. And again the only "useful" stall, the math pipe throttle, has gone down and we are memory bound.
![image](https://user-images.githubusercontent.com/3473550/90528951-77770980-e173-11ea-968c-45cf2fce35f6.png)

As in the SHARED case, again the instruction mix is very different, and we do more instructions but probably they are useless ones.. (ie some we could avoid in the ME computation)
![image](https://user-images.githubusercontent.com/3473550/90529208-c15fef80-e173-11ea-8f97-250dc3a4bdef.png)

Fianlly, this time the reduced number of registries (166 vs 176) does increase occupancy a bit, but this is completely irrelevant now as we have become memory bound
![image](https://user-images.githubusercontent.com/3473550/90529380-f2402480-e173-11ea-8036-9763d153a107.png)

**Comment by valassi** [comment]
Ok I have removed all code for SHARED and GLOBAL.

In case, this is where it was done: https://github.com/madgraph5/madgraph4gpu/commit/d1a5097318e80b827d36bc69b12ba8033acb5024

This can be closed now
